### DISCUSSION TITLE
Suggestions on document optimization

### DISCUSSION THREAD

**Discussion by suesunss** [sqlalchemy/sqlalchemy#7786] [Usage Questions]
This discussion is related to my comments in this [pr](https://github.com/sqlalchemy/sqlalchemy/pull/7689).

The comment is unrelated to the pull request, thus I would rather open a new issue. 

To begin with, I would firstly apologize for my pure complaints without concrete contexts and examples provided in the last post. 

I would say, It's more about the structure of the document that makes me confused, rather than typo or grammar. The structure of the documentation gives me the feeling that it is geared more towards an experienced SQLAlchemy user than a newcomer.

Let's take a few examples:

1. Homepage: https://www.sqlalchemy.org/

For newcomers like me, **Installation** and a **Quickstart** are more natural and make it easier to get started, instead of a lengthy **PHILOSOPHY**, as far as I am concerned. 

Let's take a look at [Flask - Homepage](https://flask.palletsprojects.com/en/2.0.x/). The first sentence gets to the point:

>> Welcome to Flask’s documentation. Get started with [Installation](https://flask.palletsprojects.com/en/2.0.x/installation/) and then get an overview with the [Quickstart](https://flask.palletsprojects.com/en/2.0.x/quickstart/). There is also a more detailed [Tutorial](https://flask.palletsprojects.com/en/2.0.x/tutorial/) that shows how to create a small but complete application with Flask. Common patterns are described in the [Patterns for Flask](https://flask.palletsprojects.com/en/2.0.x/patterns/) section. The rest of the docs describe each component of Flask in detail, with a full reference in the [API](https://flask.palletsprojects.com/en/2.0.x/api/) section.

2. Well, I finished the lengthy **PHILOSOPHY**, Where should I go next? 

I still cannot find below any links directed related to installation. Maybe I can go to [Library](https://www.sqlalchemy.org/library.html)?

That's it! I finally found the section called **REFERENCE DOCUMENTATION**, I clicked on Version 1.4 and it's until now we finally get to the documentation.

But wait, it's the **VERY BIG** Tutorials that come to my eyes in the very beginning instead of Getting Started, as Tutorials holds a large paragraph whereas Getting Started holds only a few obscure links.

Let's take a look at what Tutorials says: 

> SQLAlchemy 2.0 is functionally available as part of SQLAlchemy 1.4 and integrates Core and ORM working styles more closely than ever. The new tutorial introduces both concepts in parallel. New users and those starting new projects should start here!

I would expect some tutorials such as **SETUP**, **API**, but I do see a few terms that are difficult for newbies to understand: **CORE** and **ORM**, without given any contexts. I know what an ORM is, but I don't know what CORE means.

Finally, I returned to the homepage and found a simple explanation of CORE and ORM on the KEY FEATURES page. Only then did I finally understand the two most important concepts of this project and I was eager to see some sample code but I'd better finish this KEY FEATURE chapter first.

3. Unclear directory tree

When navigating between sections, the directory tree behaves strangely. For instance, if we end up on page [Engine](https://docs.sqlalchemy.org/en/14/tutorial/engine.html), there is no way to go back to neither [CORE](https://docs.sqlalchemy.org/en/14/core/index.html) or [ORM](https://docs.sqlalchemy.org/en/14/orm/index.html). 

4. Who is the first citizen

In document overview, it states that **Object Relational** and **SQL Expression Language** are legacy APIs. However, when we step into the specific section, such as **SQL Expression Language**, I see no statements about **LEGACY WARNING**, except for an announcement about 1.x style deprecation.

I can probably tell that 1.x contains most of this chapter, but what's exactly 1.x? Should it be stated here, especially for a newcomer of SQLAlchemy? This creates too many distractions for beginners, making it vulnerable for newcomers to miss the point. 

What's more, if we google some brief usage documentation about SQLAlchemy, many answers point to those pages, making newcomers feel that this is the latest way (**first citizen**) to use the framework, however, apparently that is not the case.

Besides the above points, there are still some points that I am missing when I went through the documentation, we can leave them for further discussion. All in all, the most obvious problem, as far as I am concerned, is that the current documentation presents a steep learning curve for beginners, especially those who are comfortable with existing solutions such as those among Java systems (e.g. Spring Data). It is normal to have a sense of strange feeling in the first beginning, especially for beginners like me, but probably SQLAlchemy can do better for those beginners, although it now has a considerable number of existing users.

**Comment by CaselIT** [comment]
Thanks for taking the time to better express your feedback.

Not a full reply, but regarding 

> Homepage: https://www.sqlalchemy.org/

that cannot be compared with the flask link, since one is the project home, the other the documentation home.
These are a more appropriate comparison link: https://palletsprojects.com/ or https://www.djangoproject.com/
That said I agree that we can improve the visibility of the documentation / get started links and/or the page organization

**Reply by suesunss** [reply]
I am more familiar with projects like flask/NoSQL Python Client (like of ElasticSearch) than sqlalchemy. I agree that SQLAlchemy has much more stuff, thus it's reasonable to be compared with Django.

**Comment by zzzeek** [comment]
> I still cannot find below any links directed related to installation. Maybe I can go to [Library](https://www.sqlalchemy.org/library.html)?
> 
> That's it! I finally found the section called **REFERENCE DOCUMENTATION**, I clicked on Version 1.4 and it's until now we finally get to the documentation.
> 
> But wait, it's the **VERY BIG** Tutorials that come to my eyes in the very beginning instead of Getting Started, as Tutorials holds a large paragraph whereas Getting Started holds only a few obscure links.

Hey there -

I dont have time to respond to all of this now, but I'm just not sure where to go with this.    There's a top-level H1 tag "Getting Started", and then directly next to it are two links named "Overview" and "Installation Guide". and a paragraph "A high level view and getting set up".  It's the first content on the page.     So, your experience was that you were looking for how to install SQLAlchemy, but you weren't able to see the phrase "Installation Guide" because there's a large paragraph in the *next* section, so you were basically confused with how to install SQLAlchemy, because you weren't able to see the link (or if you can clarify).    I'm not sure long term we're going to be able to make docs that suit every user.  I spend a **lot** of time learning to make things flashy and eye catching but at the end of the day we're a < 5 people open source project without a design firm and market research agency writing our docs.       but compare to other big projects like MySQL: https://dev.mysql.com/  and https://dev.mysql.com/doc/ , I have to read and dig for "installation" guide on their docs too, a whole lot more in fact.    

![Screenshot from 2022-03-03 09-40-49](https://user-images.githubusercontent.com/128223/156588506-14b59d46-47e4-419e-8295-f0dc3b9b5d56.png)

**Reply by zzzeek** [reply]
one thought about the homepage.   The homepage has this "Philosophy" thing that we are saying nobody cares about.

this would indicate to me a change in audience.  Because someone new to SQLAlchemy, they see the homepage, see a big boring "Philosophy" thing, it's not their thing, and they leave.  That's been the purpose of the homepage for many years, as a **selling pitch** for the product, and if it all seemed too overblown, great, they went and used Django ORM, that's great.

But here, it seems you came to SQLAlchemy **already sold** in some way, such that all the selling pitch and background stuff was all unnecessary confusing wall of text.

So who is our user ?

Years ago, it was the **unsold technologist**, who found our site and had to determine if it's somethign they wanted to get into.

Now, it seems to be the **already sold user**, who is coming to our site because...they *want* to use SQLAlchemy?   or they *have* to?

Why are you interested in SQLAlchemy?   Is this for your job, or are you a student and it's your homework?   Why do you need to learn the tool ?    There are other ORMs that are much more in the "Quick and easy" category like [PeeWee](http://docs.peewee-orm.com/en/latest/).   If SQLAlchemy is too overwhelming, why not use something simpler for beginners?   Beginners to Python / SQL / ORMs are really not our target user.

**Reply by suesunss** [reply]
> So, your experience was that you were looking for how to install SQLAlchemy, but you weren't able to see the phrase "Installation Guide" because there's a large paragraph in the _next_ section, so you were basically confused with how to install SQLAlchemy, because you weren't able to see the link (or if you can clarify). 

I am able to install SQLAlchemy, even I don't notice the link in the first place, I can still search it. I think it's more about the typesetting of the section **GET STARTED** and **Tutorials**, it gives me an impression that section **Tutorials** has much more importance than **GET STARTED** as it holds much more space.

Also, do we really need to put things like **Migration from 1.3** and **changelog** in the section *&GET STARTED**?

**Reply by zzzeek** [reply]
So I'm messing w the site right now but I have moved that stuff - if you look at https://docs.sqlalchemy.org/en/14/ (reload) Ive reorged a bit based on these general suggestions.    yes the glossary / migration / etc doesnt belong in "getting started" so that's moved.  

it's not up yet but "Download as a ZIP File" is also misleading so there will be something clearer coming up in awhile.

![Screenshot from 2022-03-03 12-00-08](https://user-images.githubusercontent.com/128223/156613754-5c511845-cb17-481e-8ee9-b03aa980a87f.png)

**Reply by suesunss** [reply]
> Why are you interested in SQLAlchemy? Is this for your job, or are you a student and it's your homework? Why do you need to learn the tool ? There are other ORMs that are much more in the "Quick and easy" category like [PeeWee](http://docs.peewee-orm.com/en/latest/). If SQLAlchemy is too overwhelming, why not use something simpler for beginners? Beginners to Python / SQL / ORMs are really not our target user.

Here I can tell my situation, I am more experienced in Java Systems (such as Spring), we choose SQLAlchemy, as you said, it's already sold to, saying, some of my colleagues who are familiar with Python. So if I want build a system using Python which needs to connect to a database, SQLAlchemy **very likely** comes to the first place as it gives us the confidence to build more robust and flexible production-grade software.

**Reply by suesunss** [reply]
> So I'm messing w the site right now but I have moved that stuff - if you look at https://docs.sqlalchemy.org/en/14/ (reload) Ive reorged a bit based on these general suggestions. yes the glossary / migration / etc doesnt belong in "getting started" so that's moved.

This section looks much more clear and **eye catching** to me.

**Reply by wu-clan** [reply]
> one thought about the homepage. The homepage has this "Philosophy" thing that we are saying nobody cares about.
> 
> this would indicate to me a change in audience. Because someone new to SQLAlchemy, they see the homepage, see a big boring "Philosophy" thing, it's not their thing, and they leave. That's been the purpose of the homepage for many years, as a **selling pitch** for the product, and if it all seemed too overblown, great, they went and used Django ORM, that's great.
> 
> But here, it seems you came to SQLAlchemy **already sold** in some way, such that all the selling pitch and background stuff was all unnecessary confusing wall of text.
> 
> So who is our user ?
> 
> Years ago, it was the **unsold technologist**, who found our site and had to determine if it's somethign they wanted to get into.
> 
> Now, it seems to be the **already sold user**, who is coming to our site because...they _want_ to use SQLAlchemy? or they _have_ to?
> 
> Why are you interested in SQLAlchemy? Is this for your job, or are you a student and it's your homework? Why do you need to learn the tool ? There are other ORMs that are much more in the "Quick and easy" category like [PeeWee](http://docs.peewee-orm.com/en/latest/). If SQLAlchemy is too overwhelming, why not use something simpler for beginners? Beginners to Python / SQL / ORMs are really not our target user.

I'm surprised at the description of the target user. As for the document, I think it is unfriendly to most users. More people want to learn how to operate or solve urgent problems. What they need is more eye-catching and easy to understand code or instructions. There are too many long speeches and many jumps in the document, It is also difficult for me to find clear CRUD tutorials and their extension methods, such as get(), scalars(), first()... they seem to be hiding. It's really a headache to read official documents. It be like wants everyone to be directly familiar with his implementation logic, which is unreasonable

**Reply by zzzeek** [reply]
Hi  -

the current tutorial is here: https://docs.sqlalchemy.org/en/14/tutorial/index.html  

the ORM CRUD part of the tutorial is here: https://docs.sqlalchemy.org/en/14/tutorial/orm_data_manipulation.html#tutorial-orm-data-manipulation 

the code in this tutorial is runnable.   The paragraphs of text are necessary in order to describe the underlying concepts that are at work, such as the identity map.  The identity map is not just "implementation logic", it affects very much how your program is going to function.  We get a significant amount of questions from frustrated users who read other tutorials which dont explain these concepts, or they glossed over them, and now they're confused because they are lacking the conceptual knowledge to know what's going on, and they end up having to spend more time learning all those concepts anyway.  

> More people want to learn how to operate or solve urgent problems. What they need is more eye-catching and easy to understand code or instructions. 

SQLAlchemy is not for people who are trying to get something done "in a hurry" with libraries they've not used before without the need to have some conceptual understanding of the library.  it's for people who are trying to write really good applications and they are taking the time to learn the tooling that will help them do so.   For users who just need to do some CRUD and forget about it, there are great alternatives like [PeeWee](http://docs.peewee-orm.com/en/latest/) and [Django ORM](https://docs.djangoproject.com/en/4.0/topics/db/queries/).     This is why the Python community has choices in software, which is a good thing.

**Reply by wu-clan** [reply]
Thank you for your reply, let me know more about the reasons for document writing. In this way, it is reasonable. Hope to see 2.0 soon

**Comment by davidism** [comment]
Hi, Flask maintainer here. It's great that you like our docs, but it's not really an actionable comparison. The things you seem to like are a complete accident in terms of intention when writing them. I agree with zzzeek's comments here and in the linked issue regarding the success of the docs over time, their intended audiences, and the inability to write for everyone.

Personally, I'm really happy with the SQLAlchemy docs. I first learned SQLAlchemy at the same time that I learned Flask, about 8 years ago. Back then, I was familiar with programming but totally new to both libraries. As a beginner to both, I wasn't confused by either, and really appreciated the detail they went into because it helped me understand how to think about my code and how I used the libraries. Now that I maintain Flask there are so many things I want to reorganize in those docs to be more like SQLAlchemy's.

I value documentation that is detailed, organized, describes clearly how to correctly understand and use features, and sets up users for long term success. The SQLAlchemy docs are a reference I use constantly because they have these qualities already.

**Comment by zzzeek** [comment]
I've added a quickstart guide at https://docs.sqlalchemy.org/en/14/orm/quickstart.html, the purpose of which is to give users a quick look at what the ORM looks like in typical use.  contains lots of links to the real tutorial.

**Reply by zzzeek** [reply]
I'm putting up sphinx-copybutton also so that we can finally copy and paste the code examples

**Reply by wu-clan** [reply]
thank you @zzzeek. You're really great, read questions carefully by everyone, i'm glad you took the advice and acted immediately, you are an example for many workers, SQLAlchemy is great!
